src/ert/ensemble_evaluator/evaluator.py

```diff
@@ -205,43 +223,41 @@ async def _failed_handler(self, events: Sequence[EnsembleFailed]) -> None:
-        with self.store_client(websocket):
-            current_snapshot_dict = self._ensemble.snapshot.to_dict()
-            event: Event = EESnapshot(
-                snapshot=current_snapshot_dict, ensemble=self.ensemble.id_
-            )
-            await websocket.send(event_to_json(event))
-
-            async for raw_msg in websocket:
-                event = event_from_json(raw_msg)
-                logger.debug(f"got message from client: {event}")
-                if type(event) is EEUserCancel:
-                    logger.debug(f"Client {websocket.remote_address} asked to cancel.")
-                    self._signal_cancel()
-
-                elif type(event) is EEUserDone:
-                    logger.debug(f"Client {websocket.remote_address} signalled done.")
-                    self.stop()
-
-    @asynccontextmanager
-    async def count_dispatcher(self) -> AsyncIterator[None]:
-        await self._dispatchers_connected.put(None)
-        yield
-        await self._dispatchers_connected.get()
-        self._dispatchers_connected.task_done()
-
-    async def handle_dispatch(self, websocket: ServerConnection) -> None:
-        async with self.count_dispatcher():
+    async def listen_for_messages(self) -> None:
+        while True:
             try:
-                async for raw_msg in websocket:
-                    try:
+                dealer, _, *frames = await self._router_socket.recv_multipart()
+                sender = dealer.decode("utf-8")
+                if sender.startswith("client"):
+                    for frame in frames:
+                        raw_msg = frame.decode("utf-8")
+                        if raw_msg == "CONNECT":
+                            self._clients_connected.add(dealer)
+                            self._clients_empty.clear()
+                            current_snapshot_dict = self._ensemble.snapshot.to_dict()
+                            event: Event = EESnapshot(
+                                snapshot=current_snapshot_dict,
+                                ensemble=self.ensemble.id_,
+                            )
+                            await self._router_socket.send_multipart(
+                                [dealer, b"", event_to_json(event).encode()]
+                            )
+                        elif raw_msg == "DISCONNECT":
+                            self._clients_connected.remove(dealer)
+                            if not self._clients_connected:
+                                self._clients_empty.set()
+                        else:
+                            event = event_from_json(raw_msg)
+                            if type(event) is EEUserCancel:
+                                logger.debug("Client asked to cancel.")
+                                self._signal_cancel()
+                            elif type(event) is EEUserDone:
+                                logger.debug("Client signalled done.")
+                                self.stop()
```

Comment:
Do we have to ACK messages from the monitor/client, or is it self acknowledging when it sees the EE stopping?

Reply:
I've decided not to do acknowledgements between the client and evaluator, since it is dispatchers that are problematic. But time will tell...

Reply:
Now `ACK` works on both dispatcher & monitor